### PR TITLE
feat!: React navigation v5

### PR DESCRIPTION
**Pull request by conrad-vanl**
**should be merged into a 2.0 or next branch**

### What does this PR do, or why is it needed?

Upgrades to `react-navigation-v5` and switches to using native navigation where possible!

ezgif com-optimize (3)

Screen Shot 2020-07-10 at 3 54 59 PM

And, while our iPad styles are still goofy as hell, this will help us pave the way to make it better by using native navigation controls available on iPad, for example iPad's modal overlays: 
Screen Shot 2020-07-10 at 4 05 27 PM

### What design trade-offs/decisions were made?

Lots of changes to the template project

## TODO:

- [ ] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [tag-issues-here]
- [ ] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android
- [ ] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

**Comment by conrad-vanl** [comment]
Companion PR: https://github.com/ApollosProject/apollos-templates/pull/73

**Comment by conrad-vanl** [comment]
@vinnyjth @asleepingpanda @redreceipt @richarddubay I would like your feedback on how to handle the upgrade to `react-navigation-v5`.

There's a significant refactor imposed in this upgrade, unfortunately, most of the changes [take place in the template app](https://github.com/ApollosProject/apollos-templates/pull/73/files). 

The main difference is a move to a component-based syntax, instead of statically defined routes, as well as a myriad of API changes:  https://reactnavigation.org/docs/upgrading-from-4.x

There is a compatibility layer available, but it doesn't support all of the API features of V5, and unfortunately it doesn't look like you can pick and choose when you use the compat layer and when you use the new syntax.

These PRs are not done yet, but the app is functioning. 

I need to:

- [ ] Go through the remaining of the upgrade guide
- [ ] Work on decreasing footprint of PR

Right now there is a footprint increase on the -template side that I believe I can help reduce. One of the benefits of V5 is a simpler API and I believe we can leverage that API to reduce the amount of owned code on the -template side. But of course, that's going to create additional changes that will slow down upgrade.

The other benefits to react-navigation-v5 are:

- Access to native navigators (use native modals, big headers on ios, etc)
- Much better performance

**Comment by asleepingpanda** [comment]
Loving everything I'm seeing so far in these screen shots. I think this is gonna be a game changer for making the apps feel more native. Quick question: how is support for "tap on icon and scroll back to the top of the content" in v5?

**Comment by conrad-vanl** [comment]
This is officially going to be put on hold until we either have a better release versioning system (so we can release it in a MAJOR version preview), or until it's officially bet on in a cycle, due to the scope of this.

**Comment by redreceipt** [comment]
ok @vinnyjth this one is a doozy 😬

File: packages/apollos-ui-analytics/package.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
     "preset": "react-native",
     "collectCoverage": true,
     "transformIgnorePatterns": [
-      "node_modules/(?!((jest-)?react-native|rn-*|react-clone-referenced-element|react-navigation|@apollosproject))"
+      "node_modules/(?!((jest-)?react-native|rn-*|react-clone-referenced-element|@react-navigation/native|@apollosproject))"
```
Comment (vinnyjth) [inline packages/apollos-ui-analytics/package.json:24]: good catch

File: packages/apollos-ui-auth/__mocks__/react-native-safe-area-context.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
```
Comment (vinnyjth) [inline packages/apollos-ui-auth/__mocks__/react-native-safe-area-context.js:1]: just in case there wasn't enough different ways to mock things ;)

File: packages/apollos-ui-connected/src/HorizontalLikedContentFeedConnected/HorizontalLikedContentFeed/index.js [hunk 3]
```diff
@@ -108,7 +108,7 @@ class HorizontalLikedContentFeed extends Component {
 
   render() {
     const { isLoading, name, navigation, content = [] } = this.props;
-
+    console.log({ navigation });
```
Comment (vinnyjth) [inline packages/apollos-ui-connected/src/HorizontalLikedContentFeedConnected/HorizontalLikedContentFeed/index.js:111]:
```suggestion
```

File: packages/apollos-ui-connected/src/UpNextButtonConnected/UpNextButtonConnected.tests.js [hunk 4]
```diff
@@ -108,63 +112,77 @@ const navigation = {
...
-      <Providers mocks={[upNextMock]}>
-        <UpNextButtonConnected
-          finishedText={"You're done forever!"}
-          continueText="Hold on cowboy, you got more to do"
-          Component={View}
-          contentId="1"
-          navigation={navigation}
-        />
-      </Providers>
+      WithReactNavigator(
```
Comment (vinnyjth) [inline packages/apollos-ui-connected/src/UpNextButtonConnected/UpNextButtonConnected.tests.js:175]: Do you think we could roll `WithReactNavigator` into `renderWithApolloData` as a default? seems like it's always needed.

File: packages/apollos-ui-kit/src/ModalView/ModalButtons.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  size: theme.sizing.baseUnit,
+  iconPadding: theme.sizing.baseUnit * 0.5, // TODO: decreases button tappability but gives us the desired "smaller button" look
+  style: {
+    backgroundColor: Color(theme.colors.text.primary).fade(0.3),
+  },
+}))(ButtonIcon);
+
+const ModalButton = (props) => <StyledButtonIcon {...props} />;
+
+export const ModalBackButton = (props) => {
```
Comment (vinnyjth) [inline packages/apollos-ui-kit/src/ModalView/ModalButtons.js:19]: Do you think we should have prop overrides on these?

**Comment by vinnyjth** [comment]
"Find my campus" button isn't working

**Comment by vinnyjth** [comment]
`getParam` doesn't seem to be a valid method on the navigation object anymore.

**Comment by vinnyjth** [comment]
Apologies! both of those above comments belong in `templates`. I'll move them over.

**Comment by redreceipt** [comment]
> Apologies! both of those above comments belong in `templates`. I'll move them over.

either way. I fixed it.

File: packages/apollos-ui-connected/src/UpNextButtonConnected/UpNextButtonConnected.tests.js (hunk 4 above)
Comment (redreceipt) [reply packages/apollos-ui-connected/src/UpNextButtonConnected/UpNextButtonConnected.tests.js:175]: the concepts seem very separate. In the test utils package that I want to create, I'm hoping to export a `<TestProvider>` component that will contain everything inside it

File: packages/apollos-ui-connected/src/RockAuthedWebView/index.js [hunk 6]
```diff
@@ -35,22 +34,18 @@ const GET_USER_COOKIE = gql`
...
         if (loading) {
           return null;
         }
-        const cookie = get(data, 'currentUser.rock.authCookie', '');
         return (
           <Browser
-            cookie={cookie}
-            url={uri}
-            modal={modal}
+            cookie={data?.currentUser?.rock?.authCookie ?? ''}
```
Comment (vinnyjth) [inline packages/apollos-ui-connected/src/RockAuthedWebView/index.js:46]: woooooooow `??` is new to me. way nicer than `||`

File: packages/apollos-ui-connected/src/SearchFeedConnected/index.js [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import React from 'react';
 import { withProps } from 'recompose';
-import { withNavigation } from 'react-navigation';
+import { withNavigation } from '@react-navigation/compat';
```
Comment (vinnyjth) [inline packages/apollos-ui-connected/src/SearchFeedConnected/index.js:3]: Can you use the `navigation` hook in this file instead of the old `withNavigation`?

File: packages/apollos-ui-kit/src/ModalView/ModalButtons.js (hunk 5 above)
Comment (redreceipt) [reply packages/apollos-ui-kit/src/ModalView/ModalButtons.js:21]: could you elaborate

File: packages/apollos-ui-connected/src/RockAuthedWebView/index.js (hunk 6 above)
Comment (redreceipt) [reply packages/apollos-ui-connected/src/RockAuthedWebView/index.js:46]: yea it's great, doesn't return false for `0` and `""`, only `null` and `undefined`

File: packages/apollos-ui-connected/src/UpNextButtonConnected/UpNextButtonConnected.tests.js (hunk 4 above)
Comment (vinnyjth) [reply packages/apollos-ui-connected/src/UpNextButtonConnected/UpNextButtonConnected.tests.js:175]: 👍 I like that even more.

File: packages/apollos-ui-kit/src/ModalView/ModalButtons.js (hunk 5 above)
Comment (vinnyjth) [reply packages/apollos-ui-kit/src/ModalView/ModalButtons.js:21]: like wrap these components w/ `withTheme` so churches can change their props easily.

File: packages/apollos-ui-connected/src/SearchFeedConnected/index.js (hunk 7 above)
Comment (redreceipt) [reply packages/apollos-ui-connected/src/SearchFeedConnected/index.js]: done

**Review by vinnyjth — APPROVED** [review]
I dig it.

File: packages/apollos-ui-kit/src/ModalView/ModalButtons.js (hunk 5 above)
Comment (redreceipt) [reply packages/apollos-ui-kit/src/ModalView/ModalButtons.js:21]: I've never done that before, I'm sure it's simple but I'd have to learn a bit. could I create a todo and do it after this is merged?
